File: Leetcode/Python/_205.py
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        self.s2t = {}
        self.t2s = {}

        for i in range(len(s)):
            if s[i] not in self.s2t and t[i] not in self.t2s:
                self.s2t[s[i]] = t[i]
                self.t2s[t[i]] = s[i]
            elif (s[i] in self.s2t and t[i] not in self.t2s) or (s[i] not in self.s2t and t[i] in self.t2s):
                return False
            if not self.s2t[s[i]] == t[i] and not self.t2s[t[i]] == s[i]:
                return False
        return True

File: Leetcode/cpp/205.cpp
class Solution {
public:
    bool isIsomorphic(string s, string t) {
        if (s.length() != t.length()) {
            return false;
        }
        std::unordered_map<char, char> s2t, t2s;

        for (int i = 0; i < s.length(); ++i) {
            char sChar = s[i];
            char tChar = t[i];
            if (!s2t.contains(sChar) && !t2s.contains(tChar)) {
                s2t[sChar] = tChar;
                t2s[tChar] = sChar;
            } else if ((!s2t.contains(sChar) && t2s.contains(tChar)) || (s2t.contains(sChar) && !t2s.contains(tChar)) || (s2t[sChar] != tChar || t2s[tChar] != sChar)) {
                return false;
            }
        }
        return true;
    }
};
